docs/proposals/spec-reconciliation.md

```diff
@@ -0,0 +1,130 @@
+    2. If the fetch succeeds:
+        1. add new specs to the datastore
+        2. delete that specs that are removed from datastore and do not have a service instance.
+        3. mark the specs that are removed from datastore and has a service instance
+ 2. During deprovision, if the instance getting deprovisioned is the last one delete the spec from datastore
+ 3. Serve the catalog request with specs that are not marked
+ 
+This will solve the problem but the task of deleting the specs is spread across time (when fetching specs and when deprovisioning)
+
+#### Approach#2
+ 1. Fetch the specs from registry
+    1. If the fetch fails mark all the specs for deletion
+    2. If the fetch succeeds,
+       1. add new specs to the datastore
+       2. mark the specs that are removed from the registry.
+ 3. Run a background task to delete all the marked specs that does not have a service instance currently provisioned. The interval of running this task can be configurable
+ 4. Serve the catalog request with specs that are not marked
+ 
+This solves the problem but introduces a new parameter of configuration. 
+
+If in future, other conditions of deleting a spec crop up, just mark it for deletion and background task will take care of deleting it safely.
+
+
+Either approach will take care of the all objectives and bugs. 
+
+### Approaches to handle untimely provision request
+
+
+To to handle a case where provision is requested after the broker has started bootstrapping but before it has completed bootstrapping, the broker can take the following approaches:
+
+#### Approach#1
+The broker always sends an error code if it is busy provisioning. OSB [spec](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#response-2) does not have an appropriate error code for busy.
+
+#### Approach#2 
+It can block the request and when the bootstrap completes:
+ 1. provision the instance if the spec is still there in datastore and return a success code
+ 2. send an error code - 400 Bad request, if the spec is deleted in the registry
+
+#### Approach #3
+It can send a 202 Accepted status, suggesting that provisioning is in progress. The service catalog will when keep polling to check if the service instance is fully provisioned. The [last operation](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#response-1) has a mechanism where broker can tell if the provision is success of failure.
+
+### Approaches to handle out of sync provision request
+
+To handle the case where service catalog sends a request to provision an instance of a spec that has been marked for deletion the broker can take the following approaches:
+
+#### Approach#1
+Successfully serves the provision request (very similar to scenario #1): This approach might lead to unexpected problems during execution.
```

Comment:
I'm not sure what it means to "successfully serve the provision request" here, but I don't think this is the correct approach. If we know the spec has been marked for deletion, IMO it should effectively not exist anymore, and therefore we should respond as if it doesn't exist.